File: Morse/C++/MorseTest.cpp
#include <gtest/gtest.h>
#include "Morse.h"


TEST(MorseEncodeTest, SingleLetterUppercase)
{
    const morse m("A");
    EXPECT_EQ(m.encode(), "._ ");
}

TEST(MorseEncodeTest, SingleLetterLowercase)
{
    const morse m("b");
    EXPECT_EQ(m.encode(), "_... ");
}

TEST(MorseEncodeTest, WordWithSpaces)
{
    const morse m("HI");
    EXPECT_EQ(m.encode(), ".... .. ");
}

TEST(MorseEncodeTest, SentenceWithSpace)
{
    const morse m("HELLO WORLD");
    EXPECT_EQ(m.encode(),
              ".... . ._.. ._.. ___ / .__ ___ ._. ._.. _.. ");
}

TEST(MorseEncodeTest, DigitsAndSymbols)
{
    const morse m("SOS 123!");
    EXPECT_EQ(m.encode(),
              "... ___ ... / .____ ..___ ...__ _._.__ ");
}

TEST(MorseDecodeTest, SingleLetter)
{
    const morse m("");
    EXPECT_EQ(m.decode("._ "), "A");
}

TEST(MorseDecodeTest, Word)
{
    const morse m("");
    EXPECT_EQ(m.decode(".... . ._.. ._.. ___ "), "HELLO");
}

TEST(MorseDecodeTest, WithSpaceSeparator)
{
    const morse m("");
    EXPECT_EQ(m.decode(".... . ._.. ._.. ___ / .__ ___ ._. ._.. _.. "),
              "HELLO WORLD");
}

TEST(MorseDecodeTest, DigitsAndSymbols)
{
    const morse m("");
    EXPECT_EQ(m.decode("... ___ ... / .____ ..___ ...__ _._.__ "),
              "SOS 123!");
}

TEST(MorseRoundTripTest, EncodeDecodeSymmetry)
{
    const std::string text = "C IS COOL!";
    const morse m(text);
    const std::string encoded = m.encode();

    const morse m2("");
    const std::string decoded = m2.decode(encoded);

    EXPECT_EQ(decoded, "C IS COOL!");
}

TEST(MorseRoundTripTest, OnlySupportedSymbols)
{
    const std::string text = "HELLO 123";
    const morse m(text);
    const std::string encoded = m.encode();

    const morse m2("");
    const std::string decoded = m2.decode(encoded);

    EXPECT_EQ(decoded, "HELLO 123");
}

File: Morse/Rust/morse_test.rs
#[cfg(test)]
mod tests {
    use crate::mouse::Morse;

    #[test]
    fn test_encode_simple_word() {
        let morse = Morse::new("SOS".to_string());
        assert_eq!(morse.encode(), "... ___ ...");
    }

    #[test]
    fn test_decode_simple_word() {
        let morse = Morse::new("".to_string());
        assert_eq!(morse.decode("... ___ ..."), "SOS");
    }

    #[test]
    fn test_case_insensitivity() {
        let morse_upper = Morse::new("HELLO".to_string());
        let morse_lower = Morse::new("hello".to_string());
        assert_eq!(morse_upper.encode(), morse_lower.encode());
    }

    #[test]
    fn test_phrase_with_space() {
        let morse = Morse::new("HELLO WORLD".to_string());
        assert_eq!(morse.encode(), ".... . ._.. ._.. ___ / .__ ___ ._. ._.. _..");
    }

    #[test]
    fn test_numbers() {
        let morse = Morse::new("123".to_string());
        assert_eq!(morse.encode(), ".____ ..___ ...__");
        assert_eq!(morse.decode(".____ ..___ ...__"), "123");
    }

    #[test]
    fn test_punctuation() {
        let morse = Morse::new("HELLO!".to_string());
        assert!(morse.encode().ends_with("_._.__"));
        assert_eq!(morse.decode(".... . ._.. ._.. ___ _._.__"), "HELLO!");
    }

    #[test]
    fn test_round_trip() {
        let text = "TEST 123!";
        let morse = Morse::new(text.to_string());
        let encoded = morse.encode();
        let decoded = morse.decode(&encoded);
        assert_eq!(decoded, text);
    }
}
